Implement tax and Create "miscellaneous" Table

arhh opened the pull request:
A bit more was done here than what the task specified but it was all in the name of "tax".

Besides implementing tax functionality:
* There's a new table called "miscellaneous" which contains records squares that are not properties (e.g. chance, tax). There are some convenience methods in `miscellaneous.py` to execute SQL queries on the database.
* A module called `check_position` was created which should ideally be run every time a player moves position. It will run through checks to determine if the player needs to pay tax/rent, open a card, go to jail, etc. Not all of its functionality is implemented yet, but adding functionality should be straight-forward.
* A function to get the id of the game that the user was added to `game.py`.
* Some more "property" convenience methods have been added to `property.py`. They include functionality to check whether a property is owned, and get a list of board positions on where property is located.

@gabystephanov Your `charge_rent` function is called from here (`check_position`), rather than from the client, if that's okay with you. The main consequence of this is that `charge_rent` isn't in `pages.py` anymore

Also, the `get_this_game()` function in `game.py` should do what @undergroundquizscene, @gabystephanov and I were discussing earlier.

@oisdk Does it makes sense to have `check_position` here hand off property purchasing to your code (as seen in the placeholder comments)?

undergroundquizscene commented on backend/backend/game.py:
```diff
@@ -188,3 +189,25 @@ def get_games(with_game_status=None):
+                this_game_id = game
+                break
+
+    return this_game_id
```
This would be more efficient if we used a dedicated SQL query for it, I think – iterating over the games to find the right one seems wasteful. Do you think that’s something to change now or later? I’m happy with either, really.

arhh replied:
Flip, yeah of course, I didn't think of that. The efficiency probably won't be noticeable with the amount of games we'll have but it will look neater. I should be able to get that done soon.

undergroundquizscene replied:
Well, the amount of games is the reason I’m easy – if it’s easy to do we might as well, but if not it’s not important yet.

undergroundquizscene commented on backend/backend/pay_tax.py:
```diff
@@ -0,0 +1,15 @@
+
+    """
+    player = Player(player_id)
+    player.balance -= tax_payable
```
I think this bit should be in a `with` block, as the player object is being updated.

arhh replied:
Yes it should. Thanks for pointing that out.

undergroundquizscene commented on backend/backend/properties.py:
```diff
@@ -253,6 +253,53 @@ def _is_in_monopoly(self):
+                           'AND property_position = %s '
+                           'AND game_id = %s;', (property_position, game_id))
+
+            return bool(cursor.rowcount > 0)
```
Do you need to call `bool` here?

arhh replied:
It was suggested by pylint as an alternative to a fully-fledged if-statement. I think it looks fairly neat to be honest.

arhh replied:
Gaby just pointed this out and now I realise this is what you probably meant: The `bool()` function didn't need to be there, instead just have `cursor.rowcount > 0`. Apologies for the mis-understanding.

undergroundquizscene replied:
No apologies needed, I should have been clearer! It hadn't occurred to me that the simpler form wouldn't have been obvious.

undergroundquizscene commented on backend/pages.py:
```diff
@@ -13,5 +13,5 @@
     'join_game': 'backend.join_game:join_game',
     'increment_turn': 'backend.increment_turn:increment_turn',
-    'charge_rent': 'backend.charge_rent:charge_rent',
+    'check_position': 'backend.check_position:check_position',
```
Does `check_position` need to be a page? I don’t think we’ll be calling it from the client.

arhh replied:
What I'm thinking is that `check_position` would be called every time a player has moved. When I think about it though, the server knows when a player has moved so it could call this module.

I'm not sure where it would be called from, though? Maybe `player_roll_dice` because that's where the server increments the players position in the database

undergroundquizscene replied:
Yeah, I’d say we could just call it from the page for rolling.

arhh replied:
What I'm thinking is that `check_position` would be called every time a player has moved. When I think about it though, the server knows when a player has moved so it could call this module.

I'm not sure where it would be called from, though?

gabystephanov commented on initialise_server.sql:
```diff
@@ -29,6 +29,13 @@ CREATE TABLE IF NOT EXISTS playing_in (
+    board_position tinyint UNSIGNED NOT NULL,
+    type ENUM('tax', 'chance', 'community_chest', 'jail', 'parking', 'to_jail') NOT NULL,
+    value smallint UNSIGNED,
+    PRIMARY KEY (board_position)
```
We could consider having a cards table holding the chance and community chest cards having a **id** as a primary key for card uniqueness, **string** col for the text displayed on the card, **type** enum if it's a chance or community chest card and **action** could represent a case number for a particular action (increase player bal. by x amount, roll again... etc.)

arhh replied:
Yeah definitely, whoever is doing Task 8.1 should take this into account.

arhh commented:
@undergroundquizscene I updated the `get_this_game` function to use an SQL query instead, let me know if it's accurate to what we were discussing earlier.

Also, I removed `check_position` from `pages.py` so hopefully it will work without issue when called with `roll_dice`.

undergroundquizscene commented on backend/backend/check_position.py:
```diff
@@ -0,0 +1,60 @@
+    """
+    # Get the player id from the client
+    details = json.load(source)
+    player_id = details["player_id"]
```
Since we’re no longer expecting this function to be called by the client, I think this should take the player id as an argument rather than loading it from stdin – does that make sense?

arhh commented:
@undergroundquizscene I just added the call to `check_position` to `player_roll_dice()`. I think it should be safe to integrate into master, what do you think? I would like to see how @oisdk has his property buying functionality designed but nevertheless, this code can be branched out from master later if needs be.

undergroundquizscene commented:
Sorry I’m so late coming back to this!

arhh commented:
That's no problem. I was making changes to it up to last night anyway
